source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1337,17 +1388,113 @@ bool ServerContextImpl::isClientEcdsaCapable(const SSL_CLIENT_HELLO* ssl_client_
+  if (SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_status_request,
+                                           &status_request_data, &status_request_len)) {
+    return true;
+  }
+
+  return false;
+}
+
+OcspStapleAction ServerContextImpl::ocspStapleAction(const ContextImpl::TlsContext& ctx,
+                                                     bool client_ocsp_capable) {
+  if (!client_ocsp_capable) {
+    return OcspStapleAction::ClientNotCapable;
+  }
+
+  auto& response = ctx.ocsp_response_;
+  if (!Runtime::runtimeFeatureEnabled("envoy.reloadable_features.check_ocsp_policy")) {
+    // Expiration check is disabled. Proceed as if the policy is LenientStapling and the response
+    // is not expired.
+    return response ? OcspStapleAction::Staple : OcspStapleAction::NoStaple;
+  }
+
+  auto policy = ocsp_staple_policy_;
+  if (ctx.is_must_staple_) {
+    // The certificate has the must-staple extension, so upgrade the policy to match.
+    policy = Ssl::ServerContextConfig::OcspStaplePolicy::MustStaple;
+  }
+
+  const bool valid_response = response && !response->isExpired();
+
+  switch (policy) {
+  case Ssl::ServerContextConfig::OcspStaplePolicy::LenientStapling:
+    if (!valid_response) {
+      return OcspStapleAction::NoStaple;
+    }
+    return OcspStapleAction::Staple;
+
+  case Ssl::ServerContextConfig::OcspStaplePolicy::StrictStapling:
+    if (valid_response) {
+      return OcspStapleAction::Staple;
+    }
+    if (response) {
+      // Expired response.
+      return OcspStapleAction::Fail;
+    }
+    return OcspStapleAction::NoStaple;
+
+  case Ssl::ServerContextConfig::OcspStaplePolicy::MustStaple:
+    if (!valid_response) {
+      return OcspStapleAction::Fail;
+    }
+    return OcspStapleAction::Staple;
+
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
 enum ssl_select_cert_result_t
 ServerContextImpl::selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello) {
   const bool client_ecdsa_capable = isClientEcdsaCapable(ssl_client_hello);
+  const bool client_ocsp_capable = isClientOcspCapable(ssl_client_hello);
+
   // Fallback on first certificate.
   const TlsContext* selected_ctx = &tls_contexts_[0];
+  auto ocsp_staple_action = ocspStapleAction(*selected_ctx, client_ocsp_capable);
   for (const auto& ctx : tls_contexts_) {
-    if (client_ecdsa_capable == ctx.is_ecdsa_) {
-      selected_ctx = &ctx;
-      break;
+    if (client_ecdsa_capable != ctx.is_ecdsa_) {
+      continue;
+    }
+
+    auto action = ocspStapleAction(ctx, client_ocsp_capable);
+    if (action == OcspStapleAction::Fail) {
+      continue;
     }
+
+    selected_ctx = &ctx;
+    ocsp_staple_action = action;
+    break;
+  }
+
+  if (client_ocsp_capable) {
+    stats_.ocsp_staple_requests_.inc();
   }
+
+  switch (ocsp_staple_action) {
+  case OcspStapleAction::Staple: {
+    // We avoid setting the OCSP response if the client didn't request it, but doing so is safe.
+    RELEASE_ASSERT(selected_ctx->ocsp_response_,
```

Comment:
Hi! I was just doing a pass on this PR and had a quick question to understand this better -- why is this a RELEASE_ASSERT? 
My read initially was that the ocsp_response_ must be non-null since for OcspStapleAction::Staple we must have set the response (or otherwise thrown an error before if it didn't match its cert). My question is, is this a program invariant? 

I was reading this in regards to envoy style: 
> Per above it's acceptable to turn failures into crash semantics via RELEASE_ASSERT(condition) or PANIC(message) if there is no other sensible behavior, e.g. in OOM (memory/FD) scenarios. Only RELEASE_ASSERT(condition) should be used to validate conditions that might be imposed by the external environment. ASSERT(condition) should be used to document (and check in debug-only builds) program invariants.